### PR TITLE
make friend function for uart callback handlers

### PR DESCRIPTION
**Pull request by alvinsunyixiao**

### CODE REVIEW

File: examples/uart/main.cc [hunk 1]
```diff
@@ -53,8 +53,9 @@ static osEvent uart_event;
 
 class CustomUART : public bsp::UART {
  public:
-  CustomUART(UART_HandleTypeDef *huart) : bsp::UART(huart) {}
+  using bsp::UART::UART;
```
Comment (AzulRadio) [inline examples/uart/main.cc:56]: ```CustomUART``` inherits ```UART``` and it should also inherit the constructor from ```UART```. Mind explaining why we are ```using``` the constructor of the parent class here?

**Review by AzulRadio — COMMENTED** [review]
The code looks good! 
@AlchemicRonin Could you do the verification on the board? Thank you! (As soon as you've verified it you can approve this pr)

File: examples/uart/main.cc (hunk 1 above)
Comment (alvinsunyixiao) [reply examples/uart/main.cc:56]: Nope, I don't think the constructor is inherited, at least not in C++11

File: shared/bsp/bsp_uart.cc [hunk 2]
```diff
@@ -92,13 +92,31 @@ static HAL_StatusTypeDef uart_receive_dma_double_buffer(
...
+void TxCompleteCallbackWrapper(UART_HandleTypeDef *huart) {
   UART *uart = find_uart_instance(huart);
   if (!uart)
     return;
   uart->TxCompleteCallback();
 }
 
+/* rx idle line detected -> trigger rx callback */
+void RxCompleteCallbackWrapper(UART_HandleTypeDef *huart) {
+  bsp::UART *uart = bsp::find_uart_instance(huart);
```
[the remark this answers is not in the document]
Comment (alvinsunyixiao) [reply shared/bsp/bsp_uart.cc]: Done

File: shared/bsp/bsp_uart.cc [hunk 3]
```diff
@@ -144,7 +162,7 @@ void UART::SetupTx(uint32_t tx_buffer_size) {
   tx_write_ = new uint8_t[tx_buffer_size];
   tx_read_ = new uint8_t[tx_buffer_size];
 
-  HAL_UART_RegisterCallback(huart_, HAL_UART_TX_COMPLETE_CB_ID, uart_tx_complete_callback);
+  HAL_UART_RegisterCallback(huart_, HAL_UART_TX_COMPLETE_CB_ID, TxCompleteCallbackWrapper);
```
[the remark this answers is not in the document]
Comment (alvinsunyixiao) [reply shared/bsp/bsp_uart.cc:165]: `XXX_RegisterCallback` is part of initialization. If you are referring to the SPI logic in #11, you can see `HAL_SPI_RegisterCallback` is called inside the constructor. 

Registering a callback basically means storing this function pointer somewhere in HAL such that the function is triggered when some specific event happens. In this case, when the event `UART_TX_COMPLETE` happens, we want `TxCompleteCallbackWrapper` to be called.

File: shared/bsp/bsp_uart.cc [hunk 4]
```diff
@@ -239,19 +257,6 @@ void UART::RxCompleteCallback() {}
...
-  if (__HAL_UART_GET_FLAG(huart, UART_FLAG_IDLE) && __HAL_UART_GET_IT_SOURCE(huart, UART_IT_IDLE)) {
-    uart->RxCompleteCallback();
-    __HAL_UART_CLEAR_IDLEFLAG(huart);
-  }
-
-  // TODO(alvin): add actual error handler in the future, ignore it for now
-  if (__HAL_UART_GET_IT_SOURCE(huart, UART_IT_ERR)) {
-    __HAL_UART_CLEAR_PEFLAG(huart);
-  }
+  bsp::RxCompleteCallbackWrapper(huart);
```
[the remark this answers is not in the document]
Comment (alvinsunyixiao) [reply shared/bsp/bsp_uart.cc:260]: Yes we do need this. `RM_UART_IRQHandler` is a weak function that needs to be overridden. And unfortunately, HAL does not support "registering" callback for the UART_IDLE_LINE event (which indicates uart rx complete when the packet length is unknown). So we have to do this in a slightly convoluted way.

**Comment by AlchemicRonin** [comment]
Good Job!

**Review by AlchemicRonin — APPROVED** [review]
Seems everything works fine.

**Pull request merged**